src/libraries/System.Memory.Data/src/System/BinaryData.cs

```diff
@@ -0,0 +1,262 @@
+        /// <returns>A value representing all of the data remaining in <paramref name="stream"/>.</returns>
+        public static Task<BinaryData> FromStreamAsync(Stream stream, CancellationToken cancellationToken = default)
+        {
+            if (stream == null)
+            {
+                throw new ArgumentNullException(nameof(stream));
+            }
+
+            return FromStreamAsync(stream, true, cancellationToken);
+        }
+
+        private static async Task<BinaryData> FromStreamAsync(Stream stream, bool async, CancellationToken cancellationToken = default)
+        {
+            int streamLength = 0;
+            if (stream.CanSeek)
+            {
+                long longLength = stream.Length - stream.Position;
+                if (longLength > int.MaxValue)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(stream), SR.ArgumentOutOfRange_StreamLengthMustBeInt32);
+                }
+                streamLength = (int)longLength;
+            }
+
+            using MemoryStream memoryStream = stream.CanSeek ? new MemoryStream(streamLength) : new MemoryStream();
+            if (async)
+            {
+                await stream.CopyToAsync(memoryStream, CopyToBufferSize, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                stream.CopyTo(memoryStream);
+            }
+            return new BinaryData(memoryStream.GetBuffer().AsMemory(0, (int)memoryStream.Position));
+        }
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance by serializing the provided object using
+        /// the <see cref="JsonSerializer"/>.
+        /// </summary>
+        ///
+        /// <typeparam name="T">The type to use when serializing the data.</typeparam>
+        /// <param name="jsonSerializable">The data to use.</param>
+        /// <param name="options">The options to use when serializing to JSON.</param>
+        ///
+        /// <returns>A value representing the UTF-8 encoding of the JSON representation of <paramref name="jsonSerializable" />.</returns>
+        public static BinaryData FromObjectAsJson<T>(T jsonSerializable, JsonSerializerOptions? options = default)
+        {
+            byte[] buffer = JsonSerializer.SerializeToUtf8Bytes(jsonSerializable, typeof(T), options);
+            return new BinaryData(buffer);
+        }
+
+        /// <summary>
+        /// Converts the value of this instance to a string using UTF-8.
+        /// </summary>
+        /// <returns>
+        /// A string from the value of this instance, using UTF-8 to decode the bytes.
+        /// </returns>
+        public override string ToString()
+        {
+            if (MemoryMarshal.TryGetArray(_bytes, out ArraySegment<byte> data))
+            {
+                return Encoding.UTF8.GetString(data.Array, data.Offset, data.Count);
+            }
+            return Encoding.UTF8.GetString(_bytes.ToArray());
```

Comment:
This is unnecessarily expensive for memories wrapping things other than arrays.  If this were targeting core, it could just be:
```C#
return Encoding.UTF8.GetString(_bytes.Span);
```
With it targeting .NET Standard 2.0, it can instead be:
```C#
ReadOnlySpan<byte> span = _bytes.Span;
fixed (byte* ptr = span)
{
    return Encoding.UTF8.GetString(ptr, span.Length);
}
```
That could be the entirety of the implementation, or if we wanted to avoid the pinning in the array case, it could just be the fallback, although many encodings are likely to pin the array anyway.

Reply:
It looks like UTF8 always pins here anyway.  Will take this suggestion.

Reply:
Looks like UTF8 is always pinning, I'll update to your suggestion.